specification/cognitiveservices/data-plane/Personalizer/preview/v1.1-preview.3/Personalizer.json

```diff
@@ -0,0 +1,1970 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "v1.1-preview.3",
+    "title": "Personalizer Client v1.1-preview.3",
+    "description": "Personalizer Service is an Azure Cognitive Service that makes it easy to target content and experiences without complex pre-analysis or cleanup of past data. Given a context and featurized content, the Personalizer Service returns which content item to show to users in rewardActionId. As rewards are sent in response to the use of rewardActionId, the reinforcement learning algorithm will improve the model and improve performance of future rank calls."
+  },
+  "basePath": "/personalizer/v1.1-preview.3",
+  "schemes": [
+    "https"
+  ],
+  "paths": {
+    "/configurations/service": {
+      "get": {
+        "tags": [
+          "ConfigurationsV1Dot1Preview1"
+        ],
+        "summary": "Get Service Configuration.",
+        "description": "Get the Personalizer service configuration.",
+        "operationId": "ServiceConfiguration_Get",
+        "consumes": [],
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [],
+        "responses": {
+          "200": {
+            "description": "Success",
+            "schema": {
+              "$ref": "#/definitions/ServiceConfiguration"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Successful ServiceConfiguration_Get request": {
+            "$ref": "./examples/ServiceConfiguration_Get.json"
+          }
+        }
+      },
+      "put": {
+        "tags": [
+          "ConfigurationsV1Dot1Preview1"
+        ],
+        "summary": "Update Service Configuration.",
+        "description": "Update the Personalizer service configuration.",
```

Comment:
Is this operation an "update" or a "replace"?  Meaning, if there are properties of ServiceConfiguration that are not present in the request body, do they retain their previous values ("update") or are the previous values erased ("replace")?  A "put" is typically a "replace", and if that is the behavior of this operation you should probably change the summary and description to make that explicit.

Reply:
In some cases where the property is optional in the ServiceConfiguration object (since this feature was added later, we had to make the property optional - it is called LearningMode), not providing it at all will retain the previous value - like an update operation. Also LearningMode is an enum that needs to a specific value, it cannot be removed even if the customer does not send us the value. IsAutoOptimizationEnabled, AutoOptimizationFrequency, AutoOptimizationStartDate are similar to this as well.
In some cases (like logMirrorSasUri), it is an optional feature that the customer can use. So not sending it would result in removing the existing property - like a replace operation.

What would you recommend in this case, @mikekistler ?

Reply:
Generally a `put` operation should be a full replace -- not an update. An update -- where current fields retain their value if not passed -- should be implemented as a `patch` operation that uses JSON merge patch semantics.  And `patch` can also be used for create, so recasting this as a `patch` operation is what I would recommend. Any concerns with that?